File: llm_eval_frontend/src/api/ollama.ts
/**
 * Ollama API クライアント
 * バックエンドの /api/v1/ollama エンドポイントと連携
 */
import { apiClient } from './client';
import { OllamaModelDownload } from '../types/ollama';

// 直接Ollamaサーバーにアクセスする場合は、proxy-ollama.tsを使用してください

// 共通のマッピング関数
const mapDownloadItem = (item: any): OllamaModelDownload => ({
  id: item.id,
  modelId: item.model_id,
  modelName: item.model_name,
  status: item.status,
  progress: item.progress || 0,
  totalSize: item.total_size || 0,
  downloadedSize: item.downloaded_size || 0,
  modelSize: item.model_size || 0,
  modelSizeGb: item.model_size_gb || 0.0,
  error: item.error,
  endpoint: item.endpoint || '',
  createdAt: item.created_at,
  updatedAt: item.updated_at,
  completedAt: item.completed_at,
  digest: item.digest,
  modelInfo: item.model_info
});

// バックエンドAPI経由でOllamaの操作を行う関数群
export const ollamaApi = {
  /**
   * Ollamaモデルのダウンロードを開始
   * @param modelId - モデルID
   * @param modelName - モデル名
   * @param endpoint - エンドポイント（オプション）
   * @returns ダウンロード情報
   */
  downloadModel: async (
    modelId: string,
    modelName: string,
    endpoint?: string
  ): Promise<OllamaModelDownload> => {
    try {
      const response = await apiClient.post<any>('/api/v1/ollama/download', {
        model_id: modelId,
        model_name: modelName,
        endpoint: endpoint
      });
      
      console.log('Ollama download model response:', response);
      
      // レスポンスがオブジェクト形式の場合
      if (response && typeof response === 'object') {
        return {
          id: response.download_id || response.id,
          modelId: response.model_id || modelId,
          modelName: response.model_name || modelName,
          status: response.status,
          progress: response.progress || 0,
          endpoint: response.endpoint || endpoint || '',
          createdAt: response.created_at || new Date().toISOString(),
          updatedAt: response.updated_at || new Date().toISOString()
        };
      }
      
      throw new Error('Ollama model download failed: Unexpected response format');
    } catch (error) {
      console.error('Error downloading Ollama model:', error);
      throw error;
    }
  },
  
  /**
   * ダウンロード情報を取得
   * @param downloadId - ダウンロードID
   * @returns ダウンロード情報
   */
  getDownloadStatus: async (downloadId: string): Promise<OllamaModelDownload> => {
    try {
      const response = await apiClient.get<any>(`/api/v1/ollama/download/${downloadId}`);

      console.log('Ollama download status response:', response);

      // レスポンスがオブジェクト形式の場合
      if (response && typeof response === 'object') {
        // サイズ値のログ出力（デバッグ用）
        console.log('サイズ情報:', {
          totalSize: response.total_size,
          downloadedSize: response.downloaded_size,
          modelSize: response.model_size,
          modelSizeGb: response.model_size_gb
        });

        // サイズ情報のサニティチェック
        let modelSize = 0;
        let modelSizeGb = 0;

        // モデルサイズが0、未定義、またはnullの場合
        if (!response.model_size || response.model_size <= 0) {
          console.log('モデルサイズが有効ではありません：', response.model_size);

          // 代わりにダウンロードサイズを使用
          if (response.total_size && response.total_size > 0) {
            console.log('ダウンロードサイズを代用:', response.total_size);
            modelSize = response.total_size;
            modelSizeGb = Number((response.total_size / (1024 * 1024 * 1024)).toFixed(2));
          }
        } else {
          modelSize = response.model_size;
          modelSizeGb = response.model_size_gb;
        }

        return {
          id: response.id,
          modelId: response.model_id,
          modelName: response.model_name,
          status: response.status,
          progress: response.progress || 0,
          totalSize: response.total_size || 0,
          downloadedSize: response.downloaded_size || 0,
          modelSize: modelSize,
          modelSizeGb: modelSizeGb,
          error: response.error,
          endpoint: response.endpoint || '',
          createdAt: response.created_at,
          updatedAt: response.updated_at,
          completedAt: response.completed_at,
          digest: response.digest,
          modelInfo: response.model_info
        };
      }

      throw new Error('Could not retrieve download status: Unexpected response format');
    } catch (error) {
      console.error('Error getting Ollama download status:', error);
      throw error;
    }
  },
  
  /**
   * モデルIDに関連するすべてのダウンロード情報を取得
   * @param modelId - モデルID
   * @returns ダウンロード情報の配列
   */
  getModelDownloads: async (modelId: string): Promise<OllamaModelDownload[]> => {
    try {
      const response = await apiClient.get<any>(`/api/v1/ollama/downloads/model/${modelId}`);
      
      console.log('Ollama model downloads response:', response);
      
      // レスポンスが配列形式の場合
      if (Array.isArray(response)) {
        return response.map(mapDownloadItem);
      }
      
      // レスポンスがオブジェクトでdownloadsプロパティがある場合
      if (response && typeof response === 'object' && 'downloads' in response && Array.isArray(response.downloads)) {
        return response.downloads.map(mapDownloadItem);
      }
      
      return [];
    } catch (error) {
      console.error('Error getting Ollama model downloads:', error);
      return [];
    }
  },
  
  /**
   * モデルが存在するかチェック
   * @param modelName - モデル名
   * @param endpoint - エンドポイント
   * @returns チェック結果
   */
  checkModelExists: async (modelName: string, endpoint?: string): Promise<{exists: boolean, modelInfo?: any, availableModels?: string[]}> => {
    try {
      const params = new URLSearchParams({
        model_name: modelName
      });
      
      if (endpoint) {
        params.append('endpoint', endpoint);
      }
      
      const response = await apiClient.get<any>(`/api/v1/ollama/check_model?${params.toString()}`);
      
      console.log('Ollama check model response:', response);
      
      return {
        exists: response.exists || false,
        modelInfo: response.model_info,
        availableModels: response.available_models
      };
    } catch (error) {
      console.error('Error checking Ollama model existence:', error);
      return { exists: false };
    }
  },
  
  /**
   * すべてのダウンロード情報を取得
   * @returns ダウンロード情報の配列
   */
  getAllDownloads: async (): Promise<OllamaModelDownload[]> => {
    try {
      const response = await apiClient.get<any>('/api/v1/ollama/downloads');
      
      console.log('Ollama all downloads response:', response);
      
      // レスポンスが配列形式の場合
      if (Array.isArray(response)) {
        return response.map(mapDownloadItem);
      }
      
      // レスポンスがオブジェクトでdownloadsプロパティがある場合
      if (response && typeof response === 'object' && 'downloads' in response && Array.isArray(response.downloads)) {
        return response.downloads.map(mapDownloadItem);
      }
      
      return [];
    } catch (error) {
      console.error('Error getting all Ollama downloads:', error);
      return [];
    }
  }
};

File: llm_eval_backend/src/app/api/endpoints/ollama.py
"""
Ollama管理APIエンドポイント

Ollamaモデルのダウンロードを管理するエンドポイントを実装します。
"""
import logging
from typing import List, Dict, Any, Optional
from fastapi import APIRouter, HTTPException, Depends, status, Request
from pydantic import BaseModel

from app.utils.ollama_manager import get_ollama_manager, DownloadStatus
from app.utils.db.models import get_model_repository
from app.utils.db.providers import get_provider_repository

# ロガーの設定
logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ollama", tags=["ollama"])


class OllamaModelDownloadRequest(BaseModel):
    """Ollamaモデルダウンロードリクエスト"""
    model_id: str
    model_name: str
    endpoint: Optional[str] = None


class OllamaModelDownloadResponse(BaseModel):
    """Ollamaモデルダウンロードレスポンス"""
    download_id: str
    model_id: str
    model_name: str
    status: DownloadStatus
    progress: int
    endpoint: str


class OllamaModelDownloadDetailResponse(BaseModel):
    """Ollamaモデルダウンロード詳細レスポンス"""
    id: str
    model_id: str
    model_name: str
    endpoint: str
    status: DownloadStatus
    progress: int
    total_size: int  # ダウンロード時の転送サイズ (バイト単位)
    downloaded_size: int  # ダウンロード済みの転送サイズ (バイト単位)
    model_size: int = 0  # モデルの実際のサイズ (バイト単位)
    model_size_gb: float = 0.0  # モデルの実際のサイズ (GB単位)
    error: Optional[str] = None
    created_at: str
    updated_at: str
    completed_at: Optional[str] = None
    digest: Optional[str] = None
    model_info: Optional[dict] = None  # モデルの詳細情報


@router.post("/download", response_model=OllamaModelDownloadResponse)
async def download_ollama_model(request: OllamaModelDownloadRequest):
    """
    Ollamaモデルのダウンロードを開始します。
    
    Args:
        request: ダウンロードリクエスト
        
    Returns:
        ダウンロード情報
    """
    model_repo = get_model_repository()
    provider_repo = get_provider_repository()
    
    try:
        # モデルが存在するか確認
        model = model_repo.get_model_by_id(request.model_id)
        if not model:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"モデルID '{request.model_id}' が見つかりません"
            )
        
        # プロバイダーがollamaかどうか確認
        provider = provider_repo.get_provider_by_id(model["provider_id"])
        if not provider or provider["type"].lower() != "ollama":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"モデル '{model['name']}' はOllamaプロバイダではありません"
            )
        
        # エンドポイントの取得（リクエスト > モデル > プロバイダの順）
        endpoint = request.endpoint
        if not endpoint:
            endpoint = model.get("endpoint")
        if not endpoint:
            endpoint = provider.get("endpoint")
        
        if not endpoint:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Ollamaエンドポイントが指定されていません"
            )
        
        # デフォルトのOllamaエンドポイント
        if endpoint == "ollama":
            endpoint = "http://localhost:11434"
            
        # エンドポイントがURLパスを含む場合は削除（ベースURLのみ使用）
        if '/api/' in endpoint:
            # /api/ より前の部分だけを取得
            endpoint = endpoint.split('/api/')[0]
        
        # プロトコルが含まれていることを確認
        if not endpoint.startswith(('http://', 'https://')):
            endpoint = f"http://{endpoint}"
        
        # Ollamaマネージャの取得とダウンロード開始
        ollama_manager = get_ollama_manager()
        download_info = await ollama_manager.download_model(
            model_name=request.model_name, 
            model_id=request.model_id,
            endpoint=endpoint
        )
        
        # レスポンスを作成
        return {
            "download_id": download_info["id"],
            "model_id": download_info["model_id"],
            "model_name": download_info["model_name"],
            "status": download_info["status"],
            "progress": download_info["progress"],
            "endpoint": download_info["endpoint"]
        }
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Ollamaモデルダウンロード開始エラー: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ollamaモデルダウンロード開始エラー: {str(e)}"
        )


@router.get("/download/{download_id}", response_model=OllamaModelDownloadDetailResponse)
async def get_download_status(download_id: str):
    """
    ダウンロードステータスを取得します。
    
    Args:
        download_id: ダウンロードID
        
    Returns:
        ダウンロード詳細情報
    """
    ollama_manager = get_ollama_manager()
    
    try:
        logger.info(f"[OLLAMA_API] ダウンロードステータス取得リクエスト: ID={download_id}")
        download_info = ollama_manager.get_download(download_id)
        
        if not download_info:
            logger.warning(f"[OLLAMA_API] ダウンロードIDが見つかりません: {download_id}")
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"ダウンロードID '{download_id}' が見つかりません"
            )
        
        # 返却前にステータス情報をログ出力
        status_value = download_info.get("status", "unknown")
        error_value = download_info.get("error")
        progress_value = download_info.get("progress", 0)
        model_name = download_info.get("model_name", "unknown")
        
        logger.info(f"[OLLAMA_API] ダウンロードステータス応答: ID={download_id}, モデル={model_name}, ステータス={status_value}, 進捗={progress_value}%, エラー={error_value}")
        
        return download_info
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"ダウンロードステータス取得エラー: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"ダウンロードステータス取得エラー: {str(e)}"
        )


@router.get("/downloads/model/{model_id}", response_model=List[OllamaModelDownloadDetailResponse])
async def get_downloads_by_model(model_id: str):
    """
    特定のモデルのダウンロード履歴を取得します。
    
    Args:
        model_id: モデルID
        
    Returns:
        ダウンロード情報のリスト
    """
    model_repo = get_model_repository()
    ollama_manager = get_ollama_manager()
    
    try:
        # モデルが存在するか確認
        model = model_repo.get_model_by_id(model_id)
        if not model:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"モデルID '{model_id}' が見つかりません"
            )
        
        # ダウンロード履歴を取得
        downloads = ollama_manager.get_downloads_by_model_id(model_id)
        return downloads
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"モデルダウンロード履歴取得エラー: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"モデルダウンロード履歴取得エラー: {str(e)}"
        )


@router.get("/downloads", response_model=List[OllamaModelDownloadDetailResponse])
async def get_all_downloads():
    """
    すべてのダウンロード履歴を取得します。
    
    Returns:
        ダウンロード情報のリスト
    """
    ollama_manager = get_ollama_manager()
    
    try:
        downloads = ollama_manager.get_all_downloads()
        return downloads
    
    except Exception as e:
        logger.error(f"ダウンロード履歴取得エラー: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"ダウンロード履歴取得エラー: {str(e)}"
        )


@router.get("/check_model", response_model=Dict[str, Any])
async def check_ollama_model(model_name: str, endpoint: Optional[str] = None):
    """
    Ollamaモデルの存在チェックを行います。
    
    Args:
        model_name: モデル名
        endpoint: Ollamaエンドポイント（オプション）
        
    Returns:
        モデル情報
    """
    import aiohttp
    import json
    
    # エンドポイントのデフォルト値
    if not endpoint:
        endpoint = "http://localhost:11434"
    
    # エンドポイントにプロトコルが含まれていなければ追加
    if not endpoint.startswith(('http://', 'https://')):
        endpoint = f"http://{endpoint}"
    
    # エンドポイントが正しい形式かチェック
    if not endpoint.endswith('/'):
        endpoint = f"{endpoint}/"
    
    api_url = f"{endpoint}api/tags"
    
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(api_url) as response:
                if response.status != 200:
                    error_text = await response.text()
                    logger.error(f"Ollamaモデル情報取得エラー: {error_text}")
                    raise HTTPException(
                        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                        detail=f"Ollamaモデル情報取得エラー: HTTP {response.status} - {error_text}"
                    )
                
                data = await response.json()
                
                # モデルが存在するか確認
                for model in data.get("models", []):
                    if model.get("name") == model_name:
                        return {
                            "exists": True,
                            "model_info": model
                        }
                
                # モデルが見つからない場合
                return {
                    "exists": False,
                    "available_models": [m.get("name") for m in data.get("models", [])]
                }
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Ollamaモデルチェックエラー: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ollamaモデルチェックエラー: {str(e)}"
        )